ts/test/session/sending/MessageQueue_test.ts

```diff
@@ -116,271 +68,310 @@ describe('MessageQueue', () => {
-          true,
-          'a device passed into sendMessageToDevices was not a PubKey'
-        );
-        expect(argsPaired.isEqual(pairedDevices[index])).to.equal(
-          true,
-          'a device passed into sendMessageToDevices did not match MessageDeviceProtocol.getAllDevices'
-        );
+        const eventPromise = TestUtils.waitForTask<RawMessage | OpenGroupMessage>(complete => {
+          messageQueueStub.events.once('success', complete);
+        });
+
+        await messageQueueStub.processPending(device);
+        await expect(eventPromise).to.be.fulfilled;
+
+        const rawMessage = await eventPromise;
+        expect(rawMessage.identifier).to.equal(message.identifier);
+      });
+
+      it('should send a fail event if something went wrong while sending', async () => {
+        hasSessionStub.resolves(true);
+        isMediumGroupStub.resolves(false);
+        sendStub.throws(new Error('failure'));
+
+        const spy = sandbox.spy();
+        messageQueueStub.events.on('fail', spy);
+
+        const device = TestUtils.generateFakePubKey();
+        const message = TestUtils.generateChatMessage();
+        await pendingMessageCache.add(device, message);
+
+        const eventPromise = TestUtils.waitForTask<[RawMessage | OpenGroupMessage, Error]>(complete => {
+          messageQueueStub.events.once('fail', (...args) => {
+            complete(args);
+          });
+        });
+
+        await messageQueueStub.processPending(device);
+        await expect(eventPromise).to.be.fulfilled;
+
+        const [rawMessage, error] = await eventPromise;
+        expect(rawMessage.identifier).to.equal(message.identifier);
+        expect(error.message).to.equal('failure');
       });
     });
   });
 
+  describe('sendUsingMultiDevice', () => {
+    it('should send the message to all the devices', async () => {
+      const devices = TestUtils.generateFakePubKeys(3);
+      sandbox.stub(MultiDeviceProtocol, 'getAllDevices').resolves(devices);
+      const stub = sandbox
+        .stub(messageQueueStub, 'sendMessageToDevices')
+        .resolves();
+
+      const message = TestUtils.generateChatMessage();
+      await messageQueueStub.sendUsingMultiDevice(devices[0], message);
+
+      const args = stub.lastCall.args as [Array<PubKey>, ContentMessage];
+      expect(args[0]).to.have.same.members(devices);
+      expect(args[1]).to.equal(message);
+    });
+  });
+
   describe('sendMessageToDevices', () => {
     it('can send to many devices', async () => {
-      const devices = TestUtils.generateFakePubKeys(10);
+      hasSessionStub.resolves(false);
+
+      const devices = TestUtils.generateFakePubKeys(5);
       const message = TestUtils.generateChatMessage();
 
-      const promise = messageQueueStub.sendMessageToDevices(devices, message);
+      await messageQueueStub.sendMessageToDevices(devices, message);
+      const promise = TestUtils.waitUntil(() => pendingMessageCache.getCache().length === devices.length);
```

Comment:
maybe in `sendMessageToDevices` we should do
```
return Promise.all(devices.map(d => this.process(d, message));
```
i think this was that was done earlier but i don't know why i undid it.